### DISCUSSION TITLE
Removing Temperature Outliers from CPC data using groupby and map for every lat, lon, and month separately

### DISCUSSION THREAD

**Discussion by bioinformike** [pydata/xarray#7728] [Q&A]
I asked a question over on SO a while [ago](https://stackoverflow.com/questions/71988301/setting-temp-outliers-by-month-at-each-latitude-longitude-pair-to-nan-with-xar) where I explained that I'm looking at the CPC tmax data which looks like what is in the screenshot below and I'm trying to remove outliers specifically looking at each longitude and latitude pair for each month. The answer was an adaptation of the code I had posted:

<img width="574" alt="explorer_17I7nxeDVA" src="https://user-images.githubusercontent.com/31513230/230232302-ab08a9d6-6548-432e-883e-c186d06a5422.png"> 

```
def simple_out(temp_dat):
    RECORD_HI  = 57
    RECORD_LOW = -90
    TUKEY_MULT = 3

    # in a N-D DataArray, this actually doesn't do anything
    # as the NaN cells can't be dropped individually, so I've
    # commented the next line out. quantile in xarray
    # skips NaNs by default so you're ok ignoring this.
    # no_na_ls = temp_dat[~np.isnan(temp_dat)]

    # Calculate median and  IQR for tukey
    qs = temp_dat.quantile([0.25, 0.5, 0.75], dim='time')
    pre_med = qs.sel(quantile=0.5)
    iqr = qs.sel(quantile=0.75) - qs.sel(quantile=0.25)

    OUTLIER_MIN_THRESH = pre
<img width="574" alt="explorer_17I7nxeDVA" src="https://user-images.githubusercontent.com/31513230/230232197-1254bd72-4939-4c56-af87-ce16201c21a1.png">
_med - (TUKEY_MULT * iqr)
    OUTLIER_MAX_THRESH = pre_med + (TUKEY_MULT * iqr)

    # apply conditions to all cells at once with DataArray.where
    return temp_dat.where(
        (temp_dat <= RECORD_HI)
        & (temp_dat >= RECORD_LOW)
        & (temp_dat >= OUTLIER_MIN_THRESH)
        & (temp_dat <= OUTLIER_MAX_THRESH)
    )
```

Then run:
`masked = da.tmax.groupby('time.month').map(simple_out)`

After looking at this some more, I'm questions if this is actually operating on each latitude and longitude separately for each month, or is this just working on each month separately across all latitudes and longitudes? 

For example, I was hoping to have this function run separately on the following:
[89.75, 0.25, January]
[89.72, 0.25, February]
...
[89.25, 0.25, January]
[89.25, 0.25, February]
...

I have tried throwing 'lat' and 'lon' in the groupby but that gives an error:

 ```
all_dat.tmax.groupby(['lat', 'lon', 'time.month']).map(func = simple_out, args = (TUKEY_MULT,))

TypeError: `group` must be an xarray.DataArray or the name of an xarray variable or dimension. Received ['lat', 'lon', 'time.month'] instead.
```
I have looked into this error a bit more, but I guess I'm a bit stuck as far as what to do if I can't include multiple terms in the groupby, which seems to be the case with xarray. Does anyone have any ideas on how I can handle this?

**Comment by dcherian** [comment]
Seems like  it. You can iterate over the groups to figure this out. See [this tutorial](https://tutorial.xarray.dev/intermediate/01-high-level-computation-patterns.html#instead-looping-over-groupby-objects-is-possible)
